nixos/modules/security/acme.nix

```diff
@@ -364,8 +387,63 @@ in
                       in
                         "+${script}";
                   };
+                };
 
+                # For certificates using the acme-dns dnsProvider, we
+                # handle registration and CNAME checking ourselves
+                # rather than letting lego do it, as it only attempts
+                # registration upon renewal, leading to unpredictable
+                # timing of the manual interventions required to add
+                # the CNAME records.
+                acmeDnsService = {
+                  description = "Register acme-dns Credentials for ${cert}";
+
+                  wants = [ "network-online.target" ];
+                  after = [ "network-online.target" ];
+
+                  serviceConfig = commonServiceConfig;
+
+                  # TODO: is openssl needed here? (needs testing with HTTPS
+                  # acme-dns API)
+                  path = [ pkgs.curl pkgs.openssl pkgs.dnsutils pkgs.jq ];
+                  script = ''
+                    set -uo pipefail
+
+                    if ! [ -e "$ACME_DNS_STORAGE_PATH" ]; then
+                      # We use --retry because the acme-dns server might
+                      # not be up when the service starts (especially if
+                      # it's local).
+                      response=$(curl --fail --silent --show-error \
+                        --request POST "$ACME_DNS_API_BASE/register" \
+                        --max-time 30 --retry 5 --retry-connrefused \
+                        | jq ${escapeShellArg "{${builtins.toJSON cert}: .}"})
+                      # Write the response. We do this separately to the
+                      # request to ensure that $ACME_DNS_STORAGE_PATH
+                      # doesn't get written to if curl or jq fail.
+                      echo "$response" > "$ACME_DNS_STORAGE_PATH"
+                    fi
+
+                    src='_acme-challenge.${cert}.'
+                    if ! target=$(jq --exit-status --raw-output \
+                        '.${builtins.toJSON cert}.fulldomain' \
+                        "$ACME_DNS_STORAGE_PATH"); then
+                      echo "$ACME_DNS_STORAGE_PATH has invalid format."
+                      echo "Try removing it and then running:"
+                      echo '  systemctl restart acme-${cert}.service'
+                      exit 1
+                    fi
+
+                    if ! dig +short CNAME "$src" | grep -qF "$target"; then
+                      echo "Required CNAME record for $src not found."
+                      echo "Please add the following DNS record:"
+                      echo "  $src CNAME $target."
+                      echo "and then run:"
+                      echo '  systemctl restart acme-${cert}.service'
+                      exit 1
+                    fi
```

Comment:
Can the registration logic moved to the lego binary or some helper binary installed if with `acme-dns`?

I'm a bit afraid of the more flaky parts here to break in funny ways, and assume more people would benefit from being able to check the DNS configuration w.r.t. acme-dns.

Reply:
lego's [goacmedns](https://github.com/cpu/goacmedns) dependency isn't packaged separately, but if it was then we could use `goacmedns-register(1)`. It would only affect 4 lines here, so I figured it wasn't worth it as the API is just an empty POST. If you think it'd be better then I can package and use it.

Unfortunately lego handles this *really* badly natively; it doesn't even bother registering until renew time (so you have to scramble to set up CNAME records at unpredictable intervals on initial migration to acme-dns certificates) and doesn't check the CNAME records at all (instead, it just prints the record unconditionally on initial registration and then fails mysteriously if you don't set it up). It would be nice if we could delegate some of this to lego, but given that we're already having to architect our own solution to forcing renewals on configuration changes, etc., I think some of this complexity is unavoidable to ensure a reasonable user experience unless lego changes its interface/operating model.

Reply:
Opened https://github.com/go-acme/lego/issues/1119, https://github.com/go-acme/lego/issues/1120.

Reply:
Please add comments to these issues into the comment at the beginning of the block.